### PR TITLE
Ports sects from tg

### LINKED ISSUE #3707: [Bounty] $20 Port TG sects
Port all pull requests dealing with chaplain sects from TG station

https://github.com/tgstation/tgstation/pull/49434 is the first one.

It's a good way to give the chaplain something to do, and more utility than simply valid hunting the entire round.

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request

Ported prs:
https://github.com/tgstation/tgstation/pull/49434
https://github.com/tgstation/tgstation/pull/49874
https://github.com/tgstation/tgstation/pull/50708
https://github.com/tgstation/tgstation/pull/53126
https://github.com/tgstation/tgstation/pull/54618

Adds sects, there are 3 of them at the moment: default (does pretty much nothing), technophile  and Ever-Burning Candle.

Default pretty much does nothing.
Technophiles consume powercells (more power more favor) and convert people to androids
Burning Candle gets favor for sacrificing burning bodies or burning candles, can make a piece of clothing fully fireproof or create 5 infinite candles.

Every map (besides donut) gets the Altar of God which allows people to perform those rituals.

In addition `isholy()` is now `holy_role`.

closes: #3707

## Why It's Good For The Game

More chaplain stuff to do.

## Changelog
:cl: ExcessiveUseOfCobblestone, Caldony, Archanial
add: Sects!
add: Technophile and Ever-burning Candle sects!
add: Every chapel has now altair of gods which allows you to create sect and perform their specfic rites or make sacrifices that appease the gods.
/:cl:

### CODE REVIEW

**Comment by NickEmbased** [comment]
based :+1:

**Comment by DatBoiTim** [comment]
Extremely based, I can't wait to play this shit and potentially add more sects.

**Comment by DatBoiTim** [comment]
Wiki changes would probably be in order

Also does the fireproofing mean firesuit fireproof or cannot be burned to destroy the item fireproof?

**Comment by Archanial** [comment]
> 
> 
> Wiki changes would probably be in order
> 
> Also does the fireproofing mean firesuit fireproof or cannot be burned to destroy the item fireproof?

I'll do them once I get green light from maintainers.

This is the same fireproofing as xenobio does.

**Comment by PowerfulBacon** [comment]
amazing

**Comment by PowerfulBacon** [comment]
I feel like there will be a lot of non-consensual sacrifices

**Comment by DatBoiTim** [comment]
> I feel like there will be a lot of non-consensual sacrifices

This is likely the issue of a braindead chaplain. You can easily get bodies to burn from morgue.

**Comment by Archanial** [comment]
> 
> 
> I feel like there will be a lot of non-consensual sacrifices

no, it takes a while to sacrifice someone, and he has to be on fire and dead.

**Comment by Archanial** [comment]
> 
> 
> > I feel like there will be a lot of non-consensual sacrifices
> 
> This is likely the issue of a braindead chaplain. You can easily get bodies to burn from morgue.

This was also designed for tg, we still have a cloner, just go clone yourself a few times.

**Comment by bloons3** [comment]
> I feel like there will be a lot of non-consensual sacrifices

It'd be the same as soul sharding someone who didn't consent, so eat a ban

**Comment by ilikesaying** [comment]
Execution via Chaplain/recycling of soulless or catatonic bodies

File: code/modules/reagents/chemistry/reagents/alcohol_reagents.dm
Comment (park66665) [inline code/modules/reagents/chemistry/reagents/alcohol_reagents.dm]:
```suggestion
	if(M.mind?.holy_role)
```

It is about mindless mobs, not non-mobs.

File: code/modules/religion/religion_sects.dm
Comment (park66665) [inline code/modules/religion/religion_sects.dm]:
```suggestion
		. = -favor //if favor = 5 and we want to subtract 10, we'll only be able to subtract 5
```

File: code/modules/religion/religion_sects.dm
Comment (park66665) [inline code/modules/religion/religion_sects.dm]:
```suggestion
	favor = clamp(favor+amount, 0, max_favor)
```

You might or might not be surprised by this, but `clamp(0,max_favor, favor+amount)` works while being blatantly wrong. Such blatantly wrong `clamp` usage actually working has lead to multiple cases of coders using this.

File: code/modules/religion/religion_sects.dm
Comment (park66665) [inline code/modules/religion/religion_sects.dm]:
```suggestion
	favor = clamp(favor+amount, 0, max_favor)
```

File: code/modules/religion/religion_sects.dm
Comment (park66665) [inline code/modules/religion/religion_sects.dm]:
```suggestion
		for(var/obj/item/bodypart/affecting as() in hurt_limbs)
```

File: code/modules/jobs/job_types/chaplain.dm
Comment (park66665) [inline code/modules/jobs/job_types/chaplain.dm]:
```suggestion
		H.mind?.holy_role = HOLY_ROLE_PRIEST
```

`?.` works even for lhs.

File: code/modules/jobs/job_types/chaplain.dm
Comment (park66665) [inline code/modules/jobs/job_types/chaplain.dm]:
```suggestion
		GLOB.religious_sect?.on_conversion(H)
```

File: code/modules/jobs/job_types/chaplain.dm
Comment (park66665) [inline code/modules/jobs/job_types/chaplain.dm]:
```suggestion
	H.mind?.holy_role = HOLY_ROLE_HIGHPRIEST
```

File: code/modules/antagonists/clock_cult/items/clock_components.dm
Comment (park66665) [inline code/modules/antagonists/clock_cult/items/clock_components.dm]:
```suggestion
	if(iscultist(user) || (user.mind?.holy_role))
```

File: code/datums/components/religious_tool.dm
Comment (park66665) [inline code/datums/components/religious_tool.dm]:
```suggestion
	after_sect_select_cb?.Invoke()
```

File: code/datums/components/religious_tool.dm
Comment (park66665) [inline code/datums/components/religious_tool.dm]:
```suggestion
	operation_flags = _flags
	force_catalyst_afterattack = _force_catalyst_afterattack
	after_sect_select_cb = _after_sect_select_cb
	SetGlobalToLocal() //attempt to connect on start in case one already exists!
```

`after_sect_select_cb` should be set before `SetGlobalToLocal`.

File: code/datums/components/religious_tool.dm
Comment (park66665) [inline code/datums/components/religious_tool.dm]:
```suggestion
		for(var/mob/living/am_i_holy_living in GLOB.player_list)
```

File: code/datums/components/religious_tool.dm
Comment (park66665) [inline code/datums/components/religious_tool.dm]:
```suggestion
		after_sect_select_cb?.Invoke()
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
	if(!LAZYLEN(ritual_invocations))
		if(do_after(user, target = user, delay = ritual_length))
			if(invoke_msg)
				user.say(invoke_msg)
			return TRUE
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
```

This check no longer is needed since length of `ritual_invocations` is already checked above.

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
	return ..()
```

I want to see this convention happen before someone gets a headache on why `on_riteuse` doesn't work.

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
	var/mob/living/carbon/human/human2borg = locate() in movable_reltool.buckled_mobs
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
		if(istype(chosen_clothing, /obj/item/clothing/suit/hooded) || istype(chosen_clothing, /obj/item/clothing/suit/space/hardsuit))
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
		return ..()
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
		chosen_sacrifice = corpse
		if(chosen_sacrifice.stat != DEAD)
			to_chat(user, "<span class='warning'>You can only sacrifice dead bodies, this one is still alive!</span>")
			chosen_sacrifice = null
			return FALSE
		if(!chosen_sacrifice.on_fire)
			to_chat(user, "<span class='warning'>This corpse needs to be on fire to be sacrificed!</span>")
			chosen_sacrifice = null
			return FALSE
```

Better safe than sorry.

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
	return ..()
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
	return ..()
```

File: code/modules/religion/rites.dm
Comment (park66665) [inline code/modules/religion/rites.dm]:
```suggestion
		if(!movable_reltool.can_buckle) //yes, if you have somehow managed to have someone buckled to something that now cannot buckle, we will still let you perform the rite!
			to_chat(user, "<span class='warning'>This rite requires a religious device that individuals can be buckled to.</span>")
			return FALSE
		to_chat(user, "<span class='warning'>This rite requires an individual to be buckled to [movable_reltool].</span>")
		return FALSE
```

**Review by park66665 — COMMENTED** [review]
I took a cursory look at the code changes. Feel free to ask questions. (24/7 availability not guaranteed) This review consists of 23 comments, so don't forget to expand the hidden comments!

File: code/modules/jobs/job_types/chaplain.dm
Comment (Archanial) [reply code/modules/jobs/job_types/chaplain.dm]: imo it's easier to read in `if` form but okay

File: code/datums/components/religious_tool.dm
Comment (Archanial) [reply code/datums/components/religious_tool.dm]: This is not needed since we are checking for `GLOB.religious_sect` already

Comment (park66665) [reply code/datums/components/religious_tool.dm]: `after_sect_select_cb` is a callback that is supplied as an argument when the component is added.

**Review by park66665 — APPROVED** [review]
All clear. This PR is good to go.

**Pull request merged**
